### PR TITLE
Load public keys from KC instead of storing them in configuration

### PR DESCRIPTION
**Pull request by alexeykazakov**
**How to test:**

Open `GET /api/token/keys` to see all the public keys available in the auth service.

**In this PR I did the following:**

1. There are new configurations: service account private key and key ID. This key will be used to sign service account tokens. Plus there is also a deprecated service account key (is not used yet) which will be used when we need to rotate the keys.
2. During TokenManager construction we now load public keys from Keycloak (we have to use KeycloakURL env var for that) and store these keys in the TokenManager.
3. We also load the service account private keys from configuration (see 1), extract the public keys from them and add them to the list of public keys in the token manager along with the Keycloak public keys.
4. So, there are some new methods in TokenManager:
- `PublicKeys()` - returns all the RSA public keys (loaded from KC and from service account private keys). Used by goa middleware to verify tokens.
- `PublicKey(id string)` - returns the public key by the ID. Used during token parsing. Every token has a "kid" (Key ID) header, so, we know what public key should be used to verify the token which is being parsed.
- `PemKeys()` - returns all the public keys in PEM-like format. Used to expose them via `/token/keys` endpoint.
- `JsonWebKeys()` - returns all the public keys in JSON Web Keys format. Used to expose them `via /token/keys` enpoint.

5. New `/token/keys` endpoint (fixes https://github.com/fabric8-services/fabric8-auth/issues/30):
- `GET /api/token/keys` or `GET /api/token/keys?format=jwk` returns a JSON in the following (JWK) format:
```
{
  "keys": [
    {
      "alg": "RS256",
      "e": "AQAB",
      "kid": "bNq-BCOR3ev-E6buGSaPrU-0SXX8whhDlmZ6geenkTE",
      "kty": "RSA",
      "n": "vQ8p-HsTMrgcsuIMoOR1LXRhynL9YAU0qoDON6PLKCpdBv0Xy_jnsPjo5DrtUOijuJcID8CR7E0hYpY9MgK5H5pDFwC4lbUVENquHEVS_E0pQSKCIzSmORcIhjYW2-wKfDOVjeudZwdFBIxJ6KpIty_aF78hlUJZuvghFVqoHQYTq_DZOmKjS-PAVLw8FKE3wa_3WU0EkpP-iovRMCkllzxqrcLPIvx-T2gkwe0bn0kTvdMOhTLTN2tuvKrFpVUxVi8RM_V8PtgdKroxnES7SyUqK8rLO830jKJzAYrByQL-sdGuSqInIY_geahQHEGTwMI0CLj6zfhpjSgCflstvw",
      "use": "sig"
    },
    {
      "alg": "RS256",
      "e": "AQAB",
      "kid": "9MLnViaRkhVj1GT9kpWUkwHIwUD-wZfUxR-3CpkE-Xs",
      "kty": "RSA",
      "n": "nwrjH5iTSErw9xUptp6QSFoUfpHUXZ-PaslYSUrpLjw1q27ODSFwmhV4-dAaTMO5chFv_kM36H3ZOyA146nwxBobS723okFaIkshRrf6qgtD6coTHlVUSBTAcwKEjNn4C9jtEpyOl-eSgxhMzRH3bwTIFlLlVMiZf7XVE7P3yuOCpqkk2rdYVSpQWQWKU-ZRywJkYcLwjEYjc70AoNpjO5QnY-Exx98E30iEdPHZpsfNhsjh9Z7IX5TrMYgz7zBTw8-niO_uq3RBaHyIhDbvenbR9Q59d88lbnEeHKgSMe2RQpFR3rxFRkc_64Rn_bMuL_ptNowPqh1P-9GjYzWmPw",
      "use": "sig"
    }
  ]
}
```
- `GET /api/token/keys?format=pem` returns a JSON in the following (PEM-like) format:
```
{
  "keys": [
    {
      "key": "MIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEAvQ8p+HsTMrgcsuIMoOR1LXRhynL9YAU0qoDON6PLKCpdBv0Xy/jnsPjo5DrtUOijuJcID8CR7E0hYpY9MgK5H5pDFwC4lbUVENquHEVS/E0pQSKCIzSmORcIhjYW2+wKfDOVjeudZwdFBIxJ6KpIty/aF78hlUJZuvghFVqoHQYTq/DZOmKjS+PAVLw8FKE3wa/3WU0EkpP+iovRMCkllzxqrcLPIvx+T2gkwe0bn0kTvdMOhTLTN2tuvKrFpVUxVi8RM/V8PtgdKroxnES7SyUqK8rLO830jKJzAYrByQL+sdGuSqInIY/geahQHEGTwMI0CLj6zfhpjSgCflstvwIDAQAB",
      "kid": "bNq-BCOR3ev-E6buGSaPrU-0SXX8whhDlmZ6geenkTE"
    },
    {
      "key": "MIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEAnwrjH5iTSErw9xUptp6QSFoUfpHUXZ+PaslYSUrpLjw1q27ODSFwmhV4+dAaTMO5chFv/kM36H3ZOyA146nwxBobS723okFaIkshRrf6qgtD6coTHlVUSBTAcwKEjNn4C9jtEpyOl+eSgxhMzRH3bwTIFlLlVMiZf7XVE7P3yuOCpqkk2rdYVSpQWQWKU+ZRywJkYcLwjEYjc70AoNpjO5QnY+Exx98E30iEdPHZpsfNhsjh9Z7IX5TrMYgz7zBTw8+niO/uq3RBaHyIhDbvenbR9Q59d88lbnEeHKgSMe2RQpFR3rxFRkc/64Rn/bMuL/ptNowPqh1P+9GjYzWmPwIDAQAB",
      "kid": "9MLnViaRkhVj1GT9kpWUkwHIwUD-wZfUxR-3CpkE-Xs"
    }
  ]
}
```
6. Token parsing code is moved from authz package to the token manager.
7. Moved all code which is used by tests only from TokenManager to test/token package.
8. Removed some dead code.
9. Remove any calls to keycloak which query 'all entitlements' from our tests (fixes https://github.com/fabric8-services/fabric8-auth/issues/13)

Part of https://github.com/fabric8-services/fabric8-auth/issues/83
Fixes https://github.com/fabric8-services/fabric8-auth/issues/29 and https://github.com/fabric8-services/fabric8-auth/issues/13 and https://github.com/fabric8-services/fabric8-auth/issues/30

### CODE REVIEW

**Comment by fabric8cd** [comment]
@alexeykazakov snapshot fabric8-auth image is available for testing.  `docker pull docker.io/fabric8/fabric8-auth:SNAPSHOT-PR-85-1`

**Comment by sbose78** [comment]
> Plus there is also a deprecated service account key (is not used yet) which will be used when we need to rotate the keys.

Is this account key something that was used previously and no longer used, or something that would be used in future -- a bit unclear on the use-case.

**Comment by alexeykazakov** [comment]
> Is this account key something that was used previously and no longer used, or something that would be used in future -- a bit unclear on the use-case.

When we retire an active private key we move it to the deprecated secret (could use the public part of it only though) and generate a new key and put it to the private key secret. The new active one will be used to sign tokens but both will be used to verify them. So, for some time both old and new tokens will be valid. Then when all tokens are updated we remove the deprecated one.

File: token/token.go [hunk 1]
```diff
@@ -1,74 +1,323 @@
...
+	ParseToken(ctx context.Context, tokenString string) (*TokenClaims, error)
+	PublicKey(kid string) *rsa.PublicKey
+	PublicKeys() []*rsa.PublicKey
+	JsonWebKeys() JsonKeys
+	PemKeys() JsonKeys
+}
+
+// PrivateKey represents an RSA private key with a Key ID
+type PrivateKey struct {
+	KID string
```
[the remark this answers is not in the document]
Comment (alexeykazakov) [reply token/token.go]: Yes, sure.

File: token/token.go [hunk 2]
```diff
@@ -1,74 +1,323 @@
...
 
-// NewManagerWithPrivateKey returns a new token Manager for handling creation of tokens with both private and pulic keys
-func NewManagerWithPrivateKey(privateKey *rsa.PrivateKey) Manager {
+// NewManagerWithPublicKey returns a new token Manager for handling tokens with the only public key
+func NewManagerWithPublicKey(id string, key *rsa.PublicKey) Manager {
 	return &tokenManager{
-		publicKey:  &privateKey.PublicKey,
-		privateKey: privateKey,
+		publicKeysMap: map[string]*rsa.PublicKey{id: key},
+		publicKeys:    []*PublicKey{{KID: id, Key: key}},
```
[the remark this answers is not in the document]
Comment (alexeykazakov) [reply token/token.go]: I wanted to guarantee the order of the tokens so we need an array. Map is used for quick access by ID.

File: token/token.go [hunk 3]
```diff
@@ -1,74 +1,323 @@
...
+		return nil, errors.New("unable to convert public keys to PEM Keys")
+	}
+	tm.pemKeys = jsonKeys
+
+	return tm, nil
 }
 
-// NewManagerWithPrivateKey returns a new token Manager for handling creation of tokens with both private and pulic keys
-func NewManagerWithPrivateKey(privateKey *rsa.PrivateKey) Manager {
+// NewManagerWithPublicKey returns a new token Manager for handling tokens with the only public key
```
[the remark this answers is not in the document]
Comment (alexeykazakov) [reply token/token.go]: I did mean "the only public key" :) This function constructs the manager with the only key which is a public key ;) Used by tests.

File: token/token.go [hunk 4]
```diff
@@ -1,74 +1,323 @@
...
+	tm := &tokenManager{
+		publicKeysMap: map[string]*rsa.PublicKey{},
 	}
+
+	keycloakKeys, err := FetchKeys(config.GetKeycloakEndpointCerts())
+	if err != nil {
+		log.Error(nil, map[string]interface{}{}, "unable to load Keycloak public keys")
+		return nil, errors.New("unable to load Keycloak public keys")
+	}
+	for _, keycloakKey := range keycloakKeys {
```
[the remark this answers is not in the document]
Comment (alexeykazakov) [reply token/token.go]: Keycloak supports multiple keys. Don't know details how to make KC verify tokens against multiple public keys yet, but KC should have something for key rotation. I hope :)
Anyway the KC REST API returns an array of the key and we will need to support multiple public keys to support key rotation.

File: token/token.go [hunk 5]
```diff
@@ -1,74 +1,323 @@
...
+	}
+	tm.serviceAccountPrivateKey = &PrivateKey{KID: kid, Key: rsaServiceAccountKey}
+	pk := &rsaServiceAccountKey.PublicKey
+	tm.publicKeysMap[kid] = pk
+	tm.publicKeys = append(tm.publicKeys, &PublicKey{KID: kid, Key: pk})
+	log.Info(nil, map[string]interface{}{
+		"kid": kid,
+	}, "Service account private key added")
+	// Extract public key from deprecated service account private key if any and add it the manager
+	key, kid = config.GetDeprecatedServiceAccountPrivateKey()
```
[the remark this answers is not in the document]
Comment (alexeykazakov) [reply token/token.go]: What do you mean?

**Pull request merged**